R/tm_t_crosstable.R

```diff
@@ -169,7 +169,21 @@ srv_t_crosstable <- function(id, data, reporter, filter_panel_api, label, x, y,
   with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
   checkmate::assert_class(data, "tdata")
   moduleServer(id, function(input, output, session) {
-    selector_list <- teal.transform::data_extract_multiple_srv(data_extract = list(x = x, y = y), datasets = data)
+    selector_list <- teal.transform::data_extract_multiple_srv(
+      data_extract = list(x = x, y = y),
+      datasets = data,
+      select_validation_rule = list(
+        x = shinyvalidate::sv_required("Please define column for row variable that is not empty."),
+        y = shinyvalidate::sv_required("Please define column for column variable that is not empty.")
+      )
+    )
+
+    iv_r <- reactive({
```

Comment:
So I think we need validation on this?
![image](https://user-images.githubusercontent.com/15201933/209329700-76edfe59-2168-4710-9274-cb96dec3bfde.png)


I think we should add this app to the roxygen?

```

d1 <- data.frame(
  id = factor(1:10),
  val = factor(sample(10, 10)),
  cat = sample(letters, 10),
  sex = sample(c("M", "F"), 10, replace = TRUE)
)
d2 <- data.frame(
  id = factor(letters[1:10]),
  fk = factor(sample(10, 10), levels = 1:10),
  val = factor(sample(10, 10)),
  cat_2 = factor(sample(letters[1:3], 10, replace = TRUE)),
  cat_3 = sample(letters[4:6], 10, replace = TRUE)
)

data <- teal_data(
  dataset("D1", d1, keys = "id"),
  dataset("D2", d2, keys = "id"),
  join_keys = join_keys(
    join_key("D1", "D2", c("id"= "fk"))
  ),
  check = FALSE
)

x_extract <- lapply(c("D1", "D2"), function(dataname) {
  data_extract_spec(
    dataname = dataname,
    select = select_spec(
      label = "Select variable:",
      choices = variable_choices(dataname, subset = function(data) {
        idx <- !vapply(data, inherits, logical(1), c("Date", "POSIXct", "POSIXlt"))
        return(names(data)[idx])
      }),
      selected = "val",
      multiple = TRUE,
      ordered = TRUE,
      fixed = FALSE
    )
  )
})

y_extract <- lapply(c("D1", "D2"), function(dataname){
  data_extract_spec(
    dataname = dataname,
    select = select_spec(
      label = "Select variable:",
      choices = variable_choices(dataname, subset = function(data) {
        idx <- vapply(data, is.factor, logical(1))
        return(names(data)[idx])
      }),
      selected = "val",
      multiple = FALSE,
      fixed = FALSE
    )
  )
})


app <- init(
  data = data,
  modules = modules(
    tm_t_crosstable(
      label = "Cross Table",
      x = x_extract,
      y = y_extract,
      basic_table_args = teal.widgets::basic_table_args(
        subtitles = "Table generated by Crosstable Module"
      )
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}

```

Reply:
@pawelru @insightsengineering/nest-core-dev @lcd2yyz @donyunardi  this example sort of epitomises what's wrong with how app developers specify inputs to modules.

As an app developer I want to say "for datasets D1 and D2 I want all columns which make sense to be included for X and Y" but in order to do that I need > 30 lines of code and I have to intuit what makes sense making sure it matches what the module developer is expecting i.e. what column data types make sense, whether multiple should be TRUE or not, what the label for the select spec should be etc. etc. of course as an app developer I'm free to put filter specs and all sorts of stuff in my d_e_s definition whether or not it makes sense for the module.

"for datasets D1 and D2 I want all columns which make sense to be included for X and Y" should be the default option and as an app developer I should not need to specify anything beyond the fact I'm using these 2 datasets. If I wanted more locked down behaviour then sure I could restrict it by saying "only columns ... from dataset D1"

Of course this is the only module where we are explicit about the merge but that's a whole other issue,..

Reply:
validation for `join_fun` added.

Reply:
> 

`teal` has a decent module developer experience. But the app developer experience is lacking and, from what I see, hasn't improved much in the past months?

Reply:
> As an app developer I want to say "for datasets D1 and D2 I want all columns which make sense to be included for X and Y" but in order to do that I need > 30 lines of code and I have to intuit what makes sense making sure it matches what the module developer is expecting i.e. what column data types make sense, whether multiple should be TRUE or not, what the label for the select spec should be etc. etc. of course as an app developer I'm free to put filter specs and all sorts of stuff in my d_e_s definition whether or not it makes sense for the module.

The time for _des_ redesign is yet to come my friend. I still have my draft saved somewhere.